### ISSUE TITLE
Intermittent abnormal data rollback

### ISSUE THREAD

**Issue by safinbot** [input-output-hk/cardano-db-sync#825]
**OS**
Your OS: Ubuntu 20.04

**Versions**
The `db-sync` version (eg `cardano-db-sync --version`): 11.0.0
PostgreSQL version: 11.5

**Build/Install Method**
The method you use to build or install `cardano-db-sync`: docker

**Run method**
The method you used to run `cardano-db-sync` (eg Nix/Docker/systemd/none): docker

**Additional context**

**Problem Report**
Please do not include screenshots or images, but instead cut and paste any relevant log messages
or errors.

```
[db-sync-node:Info:72120] [2021-09-16 02:51:02.60 UTC] insertShelleyBlock(Alonzo): epoch 290, slot 40194371, block 6249709, hash e3e1b3edea074122faec30980f67741172db84adf4d9093772efa891b811e0c5
[db-sync-node:Info:72120] [2021-09-16 02:51:07.61 UTC] insertShelleyBlock(Alonzo): epoch 290, slot 40194376, block 6249710, hash cfad6324072ee5e7c9f00cb6b833109d69eda9c1d8e17d2fd06bf0b71827f718
[db-sync-node:Info:72120] [2021-09-16 02:51:07.61 UTC] Offline pool metadata fetch: 43 results, 0 fetch errors
[db-sync-node:Error:72120] [2021-09-16 02:51:07.62 UTC] runDBThread: DbInsertException "PoolOfflineData" (SqlError {sqlState = "23503", sqlExecStatus = FatalError, sqlErrorMsg = "insert or update on table \"pool_offline_data\" violates foreign key constraint \"pool_offline_data_pmr_id_fkey\"", sqlErrorDetail = "Key (pmr_id)=(22336) is not present in table \"pool_metadata_ref\".", sqlErrorHint = ""})
[db-sync-node:Error:72115] [2021-09-16 02:51:07.62 UTC] ChainSyncWithBlocksPtcl: DbInsertException "PoolOfflineData" (SqlError {sqlState = "23503", sqlExecStatus = FatalError, sqlErrorMsg = "insert or update on table \"pool_offline_data\" violates foreign key constraint \"pool_offline_data_pmr_id_fkey\"", sqlErrorDetail = "Key (pmr_id)=(22336) is not present in table \"pool_metadata_ref\".", sqlErrorHint = ""})
[db-sync-node.Mux:Info:72112] [2021-09-16 02:51:07.62 UTC] Bearer on ConnectionId {localAddress = LocalAddress "", remoteAddress = LocalAddress "/ipc/node.socket"} event: State: Dead
[db-sync-node.Mux:Notice:72112] [2021-09-16 02:51:07.62 UTC] Bearer on ConnectionId {localAddress = LocalAddress "", remoteAddress = LocalAddress "/ipc/node.socket"} event: Miniprotocol MiniProtocolNum 5 InitiatorDir terminated with exception DbInsertException "PoolOfflineData" (SqlError {sqlState = "23503", sqlExecStatus = FatalError, sqlErrorMsg = "insert or update on table \"pool_offline_data\" violates foreign key constraint \"pool_offline_data_pmr_id_fkey\"", sqlErrorDetail = "Key (pmr_id)=(22336) is not present in table \"pool_metadata_ref\".", sqlErrorHint = ""})
[db-sync-node.Subscription:Error:72111] [2021-09-16 02:51:07.62 UTC] Identity Application Exception: LocalAddress "/node-ipc/node.socket" DbInsertException "PoolOfflineData" (SqlError {sqlState = "23503", sqlExecStatus = FatalError, sqlErrorMsg = "insert or update on table \"pool_offline_data\" violates foreign key constraint \"pool_offline_data_pmr_id_fkey\"", sqlErrorDetail = "Key (pmr_id)=(22336) is not present in table \"pool_metadata_ref\".", sqlErrorHint = ""})
[db-sync-node.ErrorPolicy:Error:6] [2021-09-16 02:51:07.62 UTC] IP LocalAddress "/node-ipc/node.socket" ErrorPolicyUnhandledApplicationException (DbInsertException "PoolOfflineData" (SqlError {sqlState = "23503", sqlExecStatus = FatalError, sqlErrorMsg = "insert or update on table \"pool_offline_data\" violates foreign key constraint \"pool_offline_data_pmr_id_fkey\"", sqlErrorDetail = "Key (pmr_id)=(22336) is not present in table \"pool_metadata_ref\".", sqlErrorHint = ""}))
[db-sync-node.Subscription:Notice:65] [2021-09-16 02:51:08.62 UTC] Identity Restarting Subscription after 66552.655248783s desired valency 1 current valency 0
[db-sync-node.Subscription:Notice:65] [2021-09-16 02:51:08.62 UTC] Identity Starting Subscription Worker, valency 1
[db-sync-node.Subscription:Notice:114595] [2021-09-16 02:51:08.62 UTC] Identity Connection Attempt Start, destination LocalAddress "/node-ipc/node.socket"
[db-sync-node.Subscription:Notice:114595] [2021-09-16 02:51:08.62 UTC] Identity Connection Attempt End, destination LocalAddress "/node-ipc/node.socket" outcome: ConnectSuccessLast
[db-sync-node.Handshake:Info:114595] [2021-09-16 02:51:08.62 UTC] WithMuxBearer (ConnectionId {localAddress = LocalAddress "", remoteAddress = LocalAddress "/ipc/node.socket"}) Send (ClientAgency TokPropose,MsgProposeVersions (fromList [(NodeToClientV_1,TInt 764824073),(NodeToClientV_2,TInt 764824073),(NodeToClientV_3,TInt 764824073),(NodeToClientV_4,TInt 764824073),(NodeToClientV_5,TInt 764824073),(NodeToClientV_6,TInt 764824073),(NodeToClientV_7,TInt 764824073),(NodeToClientV_8,TInt 764824073),(NodeToClientV_9,TInt 764824073)]))
[db-sync-node.Handshake:Info:114595] [2021-09-16 02:51:08.62 UTC] WithMuxBearer (ConnectionId {localAddress = LocalAddress "", remoteAddress = LocalAddress "/ipc/node.socket"}) Recv (ServerAgency TokConfirm,MsgAcceptVersion NodeToClientV_9 (TInt 764824073))
[db-sync-node.Mux:Info:114595] [2021-09-16 02:51:08.62 UTC] Bearer on ConnectionId {localAddress = LocalAddress "", remoteAddress = LocalAddress "/ipc/node.socket"} event: Handshake Client end, duration 0.000414651s
[db-sync-node.Mux:Info:114596] [2021-09-16 02:51:08.62 UTC] Bearer on ConnectionId {localAddress = LocalAddress "", remoteAddress = LocalAddress "/ipc/node.socket"} event: State: Mature
[db-sync-node:Info:114599] [2021-09-16 02:51:08.62 UTC] Starting chainSyncClient
[db-sync-node:Info:114599] [2021-09-16 02:51:08.63 UTC] Cardano.Db tip is at slot 40194371, block 6249709
[db-sync-node:Info:114604] [2021-09-16 02:51:08.63 UTC] Running DB thread
[db-sync-node:Info:114604] [2021-09-16 02:51:08.64 UTC] Rolling back to slot 39916796, hash e72579ff89dc9ed325b723a33624b596c08141c7bd573ecfff56a1f7229e4d09
[db-sync-node.Subscription:Notice:65] [2021-09-16 02:51:08.65 UTC] Identity Required subscriptions started
[db-sync-node:Info:114604] [2021-09-16 02:51:08.65 UTC] Deleting 13650 blocks up to slot 40194371
[db-sync-node:Info:114604] [2021-09-16 03:12:19.47 UTC] Blocks deleted
[db-sync-node:Info:114604] [2021-09-16 03:12:30.56 UTC] Found snapshot file for slot 39916796, hash e72579ff89dc9ed325b723a33624b596c08141c7bd573ecfff56a1f7229e4d09
[db-sync-node:Info:114604] [2021-09-16 03:12:53.82 UTC] File /data/39916796-e72579ff89-289.lstate exists
[db-sync-node:Info:114604] [2021-09-16 03:15:12.19 UTC] Offline pool metadata fetch: 0 results, 57 fetch errors
[db-sync-node:Info:114604] [2021-09-16 03:19:14.31 UTC] insertShelleyBlock(Alonzo): epoch 290, slot 39997813, block 6240000, hash 9a8c7a8d4801dee30a479374f7e9b546168818590e0648c896dfeaadcc60d946
[db-sync-node:Info:114604] [2021-09-16 03:19:14.31 UTC] Offline pool metadata fetch: 41 results, 59 fetch errors
[db-sync-node:Info:114604] [2021-09-16 03:22:19.98 UTC] Offline pool metadata fetch: 42 results, 58 fetch errors
```

**Comment by jun0tpyrc** [comment]
same , occurred today once too

**Comment by rdlrt** [comment]
Your dbsync instance restarted (when you see the rollback). The reason for crash was fixed in 11.0.1

**Comment by erikd** [comment]
Rolling back to genesis is issue #810 which is fixed in the soon to be released `11.0.2` release (on the `release/11.0.x` branch). 

If you were not running it in docker it would simply be rolling back to the start of the current epoch. The cause of the rollback is
```
unDBThread: DbInsertException "PoolOfflineData" (SqlError {sqlState = "23503", sqlExecStatus = FatalError, sqlErrorMsg = "insert or update on table \"pool_offline_data\" violates foreign key constraint \"pool_offline_data_pmr_id_fkey\"", sqlErrorDetail = "Key (pmr_id)=(22336) is not present in table \"pool_metadata_ref\".", sqlErrorHint = ""})
```
which is also fixed in the `11.0.2` release (issue #806, #823).

@rdlrt for ... reasons ... we skipped the `11.0.1` release and the next release will be `11.0.2`.

**Comment by jun0tpyrc** [comment]
For  v11.0.2 compared to v11.0.0 , it seems has problem taking the `--state-dir` param
https://github.com/input-output-hk/cardano-graphql/blob/master/docker-compose.yml#L51
```
Connecting to network specified in configuration.yaml
Invalid option `--state-dir'

Usage: cardano-db-sync-extended (COMMAND | --config FILEPATH
                                  --socket-path FILEPATH --state-dir FILEPATH
                                  --schema-dir FILEPATH
                                  [--rollback-to-slot WORD])
  Extended Cardano POstgreSQL sync node.
```

**Comment by jun0tpyrc** [comment]
@erikd any guidance or suggestion, seem 11.0.0 to 11.0.2 itself not being seamless 🙏 

Even use workingdir workaround as replacement of state-dir
https://github.com/input-output-hk/cardano-db-sync/commit/ea6e191b6b5e5abd63528cef6d3faca2b249cfe2#diff-e45e45baeda1c1e73482975a664062aa56f20c03dd9d64a827aba57775bed0d3R39

it seems the new version would want to delete all blocks

**Comment by erikd** [comment]
:sob: It seems the docker stuff (which I have little knowledge of) is wrong. Reopening this.

**Comment by jbgi** [comment]
@jun0tpyrc In order the properly support snapshot restoration, the state-dir has been fixed in the docker image and not available for the user anymore.
Can you try the docker-compose file from this PR: https://github.com/input-output-hk/cardano-graphql/pull/593

**Comment by 1000101** [comment]
Same issue, out of blue when restarting [latest version ](https://github.com/input-output-hk/cardano-db-sync/commit/1b3c60514011e184b6493b6297843c143cb870f7) of `cardano-db-sync` service. Everything was running fine for almost a day until I decided to restart the service.

Running on bare metal (no docker) Linux (NixOS).

Log:

```
[db-sync-node.Handshake:Info:65] [2021-09-16 19:58:04.67 UTC] WithMuxBearer (ConnectionId {localAddress = LocalAddress "", remoteAddress = LocalAddress "/run/cardano-node/node.socket"}) Send (ClientAgency TokPropose,MsgProposeVersions (fromList [(NodeToClientV_1,TInt 764824073),(NodeToClientV_2,TInt 764824073),(NodeToClientV_3,TInt 764824073),(NodeToClientV_4,TInt 764824073),(NodeToClientV_5,TInt 764824073),(NodeToClientV_6,TInt 764824073),(NodeToClientV_7,TInt 764824073),(NodeToClientV_8,TInt 764824073),(NodeToClientV_9,TInt 764824073)]))
[db-sync-node.Handshake:Info:65] [2021-09-16 19:58:04.67 UTC] WithMuxBearer (ConnectionId {localAddress = LocalAddress "", remoteAddress = LocalAddress "/run/cardano-node/node.socket"}) Recv (ServerAgency TokConfirm,MsgAcceptVersion NodeToClientV_9 (TInt 764824073))
[db-sync-node.Mux:Info:65] [2021-09-16 19:58:04.67 UTC] Bearer on ConnectionId {localAddress = LocalAddress "", remoteAddress = LocalAddress "/run/cardano-node/node.socket"} event: Handshake Client end, duration 0.000442916s
[db-sync-node.Mux:Info:67] [2021-09-16 19:58:04.67 UTC] Bearer on ConnectionId {localAddress = LocalAddress "", remoteAddress = LocalAddress "/run/cardano-node/node.socket"} event: State: Mature
[db-sync-node:Info:70] [2021-09-16 19:58:04.67 UTC] Starting chainSyncClient
[db-sync-node:Info:70] [2021-09-16 19:58:04.69 UTC] Cardano.Db tip is at slot 40255323, block 6252757
[db-sync-node:Info:75] [2021-09-16 19:58:04.69 UTC] Running DB thread
[db-sync-node:Info:75] [2021-09-16 19:58:04.69 UTC] Rolling back to slot 39916796, hash e72579ff89dc9ed325b723a33624b596c08141c7bd573ecfff56a1f7229e4d09
[db-sync-node.Subscription:Notice:64] [2021-09-16 19:58:04.70 UTC] Identity Required subscriptions started
[db-sync-node:Info:75] [2021-09-16 19:58:04.72 UTC] Deleting 16698 blocks up to slot 40255323
```

**Comment by jbgi** [comment]
@1000101 from my understanding, filesystem checkpoints are only done at epoch boundaries, so any restart will rollback to the beginning of current epoch (which is exactly what your logs indicate).
To avoid this, for current version, I guess required restart should be done rather in the beginning of a new epoch (first day), rather than later (4th day in your case).

But this indicate we either should implement a way to rollback the DB faster, and/or find a way to do fs check-pointing more often than every epoch.

**Comment by 1000101** [comment]
> @1000101 from my understanding, filesystem checkpoints are only done at epoch boundaries, so any restart will rollback to the beginning of current epoch (which is exactly what your logs indicate).
> To avoid this, for current version, I guess required restart should be done rather in the beginning of a new epoch (first day), rather than later (4th day in your case).
> 
> But this indicates we either should implement a way to rollback the DB faster, and/or find a way to do fs check-pointing more often than every epoch.

Ah, so this is absolutely expected. Thanks for the explanation. I thought checkpoints are created every so often.

**Comment by erikd** [comment]
> But this indicates we either should implement a way to rollback the DB faster, 

Currently rollback speed is totally determined by PostgreSQL. `db-sync` issues a single `delete` command and PostgreSQL propagates this as a `delete cascade`.

> and/or find a way to do fs check-pointing more often than every epoch.

Rolling back to the start of the epoch is required because there is volatile data that is in an in-memory queue that will be lost if we don't roll back to the start of the epoch.

**Comment by vsubhuman** [comment]
> Rolling back to the start of the epoch is required because there is volatile data that is in an in-memory queue that will be lost if we don't roll back to the start of the epoch.

How big is that in-memory state? Would that be possible for the db-sync to just dump that in-memory state into a file every 1000 blocks or something like that during an epoch so then only having to roll back to the latest thousand at least?

**Comment by erikd** [comment]
It might be possible to do something like that, but it makes things more complicated (which is bad) and I am much more concerned about correctness than I am about things like this.

**Comment by erikd** [comment]
> Would that be possible for the db-sync to just dump that in-memory state into a file every 1000 blocks

Much better and easier to just prevent the rollback to begin with.

**Comment by erikd** [comment]
This ticket conflates two issues;

* An unhandled exception which has been fixed in 20ce4624f59b17151a5192d1e3bed86f85500cbc .
* A rollback to the start of the epoch on all restarts. This issue is at least partially fixed in #840 .
